opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1248,42 +1227,73 @@ namespace Opm {
 
-            prepareWellsBeforeAssembling(dt, local_deferredLogger);
+            prepareWellsBeforeAssembling(dt, deferred_logger);
         }
-        OPM_END_PARALLEL_TRY_CATCH_LOG(local_deferredLogger,
+        OPM_END_PARALLEL_TRY_CATCH_LOG(deferred_logger,
                                        "updateWellControlsAndNetworkIteration() failed: ",
-                                       this->terminal_output_, grid().comm());
+                                       this->terminal_output_, comm);
 
         // update guide rates
-        const int reportStepIdx = simulator_.episodeIndex();
         if (alq_updated || BlackoilWellModelGuideRates(*this).
-                              guideRateUpdateIsNeeded(reportStepIdx)) {
+                              guideRateUpdateIsNeeded(episodeIdx)) {
             const double simulationTime = simulator_.time();
-            const auto& comm = simulator_.vanguard().grid().comm();
             const auto& summaryState = simulator_.vanguard().summaryState();
             std::vector<Scalar> pot(this->numPhases(), 0.0);
-            const Group& fieldGroup = this->schedule().getGroup("FIELD", reportStepIdx);
+            const Group& fieldGroup = this->schedule().getGroup("FIELD", episodeIdx);
             WellGroupHelpers<Scalar>::updateGuideRates(fieldGroup,
                                                        this->schedule(),
                                                        summaryState,
                                                        this->phase_usage_,
-                                                       reportStepIdx,
+                                                       episodeIdx,
                                                        simulationTime,
                                                        this->wellState(),
                                                        this->groupState(),
                                                        comm,
                                                        &this->guideRate_,
                                                        pot,
-                                                       local_deferredLogger);
+                                                       deferred_logger);
         }
+
         // we need to re-iterate the network when the well group controls changed or gaslift/alq is changed or
         // the inner iterations are did not converge
-        const int iterationIdx = simulator_.model().newtonMethod().numIterations();
-        const bool more_network_update = this->shouldBalanceNetwork(reportStepIdx, iterationIdx) &&
-                    (more_inner_network_update || well_group_control_changed || alq_updated);
-        return {more_network_update, well_group_control_changed};
+        // but after certain number of the iterations, we terminate
+        const std::size_t max_iteration = param_.network_max_outer_iterations_;
+        if (network_update_iteration + 1 == max_iteration) {
```

Comment:
I like the original way that this type of process is in the general iteration workflow.  Basically we reach the maximum allowed number of iterations, then we do some process. It is part of the processing of the iteration process. 

The code in the current way is putting this process inside the iteration function itself, it is harder to follow. 

If you want to keep the function `updateWellControlsAndNetwork` tider, you can wrap this process in a function.

Reply:
The motivation for this refactoring is to allow for more fine-grained checks for convergence of the outer network iterations as well as more informative output. See PR https://github.com/OPM/opm-simulators/pull/6100 where part of the same issue is addressed. That is also why I would like to move the iteration workflow inside the function.